### PR TITLE
Responsive plots

### LINKED ISSUE #2678: Make plots responsive
Add a flag so that plots are responsive in a browser window - that is they resize when the window resizes.

This is something I've done manually on a number of bokeh apps e.g. https://github.com/birdsarah/gtimelog-viz/blob/master/app/templates/js/resize_plots.js

I propose that the flag is off by default, to keep the current behavior consistent.

### PR DESCRIPTION
**Pull request by birdsarah**
Fixes #2678 

At the moment this only responds to window resize events. This covers lots of useful scenarios, but leaves some out. 

Would be great if we could respond to a resize on the event on the div, and I found this useful looking little library - https://github.com/sdecima/javascript-detect-element-resize - although its looking somewhat unmaintained. 

Regardless, I believe this PR comfortably covers the feature request.

### CODE REVIEW

File: examples/embed/embed_multiple_responsive.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    {{ plot_div.green }}
+    <h3>Blue - pan no responsive</h3>
+    {{ plot_div.blue }}
+
+    {{ plot_script }}
+    </body>
+</html>
+''')
+
+resources = Resources(mode='relative-dev')
```
Comment (mattpap) [inline examples/embed/embed_multiple_responsive.py]: Is setting the `mode` necessary? Also, `relative-dev` should be only used for development.

Comment (birdsarah) [inline examples/embed/embed_multiple_responsive.py]: Ooops, that's a hangover - thanks!

**Comment by mattpap** [comment]
@birdsarah, after a cursory review, I don't see anything specifically wrong (besides resource mode, which was fixed already), but unfortunately I don't have time right now to go into more detail.

**Comment by birdsarah** [comment]
thanks for looking @mattpap - high praise indeed :wink:

**Comment by birdsarah** [comment]
The last build passed, I just restarted it to make sure it passes consistently, I'm a little worried about these tests not being robust.

**Comment by birdsarah** [comment]
OK, I've run the last build twice so feeling better about the selenium test stability.

I tested the responsive docs in the plots and its not responsive. For good reason - the container doesn't change size - which is actually how we want responsive to behave. 

I have added a warning about using with HBox, which definitely doesn't work, and that there might be other cases where it doesn't work at the moment.

File: bokeh/models/plots.py [hunk 2]
```diff
@@ -483,12 +484,22 @@ def _check_colon_in_category_label(self):
...
     support this will render via WebGL instead of the 2D canvas.
     """)
 
+    responsive = Bool(False, help="""
+    If True, the plot will automatically resize based on the size of its container. The
+    aspect ratio of the plot will be preserved, but ``plot_width`` and ``plot_height`` will
+    act only to set the initial aspect ratio.
+    .. warning::
+
+       The responsive settings is known not to work with HBox layout and may not work
```
Comment (almarklein) [inline bokeh/models/plots.py]: typo: setting -s

File: bokehjs/src/coffee/common/plot.coffee [hunk 3]
```diff
@@ -122,6 +123,11 @@ class PlotView extends ContinuumView
 
     @update_dataranges()
 
+    if @mget('responsive')
+      throttled_resize = _.throttle(@resize, 100)
+      $(window).on("resize", throttled_resize)
+      $(@resize)
```
Comment (almarklein) [inline bokehjs/src/coffee/common/plot.coffee]: Technically, the container could change size also without the window changing size. But ok for now, I think we can just hook into Phosphor resize messages/signals at some point.

Comment (birdsarah) [inline bokehjs/src/coffee/common/plot.coffee]: Absolutely. See note at start of PR. Can't wait for this to do more. If someone was feeling sneaky in the short-run and had the ability to know when their container resized they could always to a `$(window).trigger('resize')` :D

File: bokeh/models/plots.py (hunk 2 above)
Comment (birdsarah) [inline bokeh/models/plots.py]: Fixed - thanks!

**Pull request merged**